### PR TITLE
General Improvements and Resiliency

### PR DESCRIPTION
**Pull request by sheridancbio**
- Performance improvement by reading request from file rather than stdin
- Stream output to response rather than writing to file
- Add threads for processing both stdout and stderr during run of vep command line tool process
- Improved process control handling
- Add timeout for requests, so that running vep command line tool will terminate at timeout
- Encapsulate logic for wrapping one line records from command line tool into JSON list
- Encapsulate logic for emitting complete records from vep comand line tool (via buffering)

### CODE REVIEW

**Comment by sheridancbio** [comment]
This still needs polishing and squashing, but PR created for discussion and review.

File: pom.xml [hunk 1]
```diff
@@ -1,74 +1,91 @@
...
+        <!--pluginManagement-->
+            <plugins>
+                <plugin>
+                    <groupId>org.springframework.boot</groupId>
+                    <artifactId>spring-boot-maven-plugin</artifactId>
+                </plugin>
 
+                <!--plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-compiler-plugin</artifactId>
```
Comment (ao508) [inline pom.xml]: does this need to be added back in? (uncommented?)

File: src/main/java/org/genomenexus/vep_wrapper/LinesToJSONListFilterOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    /** Creates an output stream filter built on top of the specified underlying output stream
+    */
+    public LinesToJSONListFilterOutputStream(OutputStream out) {
+        super(out);
+    }
+ 
+    /** Outputs the ending JSON list bracket
+    */
+    public void	complete() throws IOException {
```
Comment (ao508) [inline src/main/java/org/genomenexus/vep_wrapper/LinesToJSONListFilterOutputStream.java]: looks like a tab was inserted here before the method name?

Comment (sheridancbio) [reply src/main/java/org/genomenexus/vep_wrapper/LinesToJSONListFilterOutputStream.java]: fixed

File: pom.xml (hunk 1 above)
Comment (sheridancbio) [reply pom.xml]: fixed

File: src/main/java/org/genomenexus/vep_wrapper/LinesToJSONListFilterOutputStream.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    public void complete() throws IOException {
+        if (out != null) {
+            if (!someDataHasBeenWritten) {
+                out.write('[');
+                lastWriteEndedLineWithoutBeginningAnother = true;
+            }
+            if (lastWriteEndedLineWithoutBeginningAnother) {
+                out.write('\n'); // write the newline which was held back to determine if a comma was needed
+            }
+            out.write(']');
```
Comment (ao508) [inline src/main/java/org/genomenexus/vep_wrapper/LinesToJSONListFilterOutputStream.java]: i like these these 2 writes are done separately for clarity.

**Comment by sheridancbio** [comment]
Today @ao508 and I tested the response for a handler which flushes part of the response through response.getOutputStream() ... and then flushes additional response body. The test demonstrated that ServletRequest is able to stream output, one chunk at time, to the outputstream without "closing / committing" the response. So, we are going to support the writing of incremental records during the processing of the vep command line tool.

Test code was:
`            out = response.getOutputStream();
            response.setContentType("text/plain");
            byte[] line1 = new byte[4];
            line1[0] = 't';  line1[1] = 'e'; line1[2] = 's'; line1[3] = 't';
            byte[] line2 = new byte[4];
            line2[0] = 'd'; line2[1] = 'o'; line2[2] = 'n'; line2[3] = 'e';
            out.write(line1);
            out.flush();
            try {Thread.sleep(60000);} catch (InterruptedException e) {}
            out.write(line2);
            out.flush();

We saw "test" appear, then a pause, then "done" appear in the output of a curl command (when using the --no-buffer flag)

File: src/main/java/org/genomenexus/vep_wrapper/VepRunner.java [hunk 4]
```diff
@@ -1,125 +1,239 @@
...
-            out.append(line);
-            previous = line;
-        }
-        if (convertToListJSON) {
-            out.append(']');
-            out.append('\n');
+            printWithTimestamp("interrupting transferrer threads");
+            vepOutputTransferrer.interrupt();
+            vepErrorTransferrer.interrupt();
+            // TODO: when the destroy of the vep process is made to destory all forked processes as well as the parent process, the code below can be tested/used:
```
Comment (averyniceday) [inline src/main/java/org/genomenexus/vep_wrapper/VepRunner.java]: leftover from dev - cleanup?

Comment (ao508) [reply src/main/java/org/genomenexus/vep_wrapper/VepRunner.java]: I think it's actually okay to leave this TODO here. It sort of outlines what we want to accomplish - if it's removed then I suggest creating an issue to address later with this information.

**Pull request merged**